SConstruct

```diff
@@ -112,6 +112,8 @@ opts.Add(
         map=architecture_aliases,
     )
 )
+opts.Add("defines", "A comma-separated list of defines to pass to the compiler, in the form of NAME or NAME=VALUE.")
```

Comment:
For consistency with the main repo's `SConstruct`, we may better expose `CCFLAGS` here. (Too bad it doesn't expose `CPPDEFINES`, which would be more explicit.)

Reply:
This probably should be done in a similar way to the main repo:

https://github.com/godotengine/godot/blob/89cc635c0554cb2e518c830969ca4c5eedda0f4e/SConstruct#L509-L524

Reply:
The reason I didn't use `CCFLAGS` is because those are compiler-specific flags, while `CPPDEFINES` is a higher abstraction list of compiler-independent flags, and scons takes care of adding the correct prefix for whatever compiler is going to be used. It makes my job much easier if I just pass a define and let scons deal with both the compiler-specific syntax and any deduplication that scons thinks needs to happen, instead of injecting things directly into `CCFLAGS`.

Reply:
That makes sense, but then I'd suggest adding support passing `CPPDEFINES` from both the command line and the environment in the same way that `CCFLAGS` etc. are supported upstream.

i.e. I'd add this upstream:
```diff
diff --git a/SConstruct b/SConstruct
index c7b9d5bc86..20c628adc2 100644
--- a/SConstruct
+++ b/SConstruct
@@ -262,6 +262,7 @@ opts.Add(BoolVariable("builtin_zstd", "Use the built-in Zstd library", True))
 opts.Add("CXX", "C++ compiler")
 opts.Add("CC", "C compiler")
 opts.Add("LINK", "Linker")
+opts.Add("CPPDEFINES", "Custom defines fo the pre-processor")
 opts.Add("CCFLAGS", "Custom flags for both the C and C++ compilers")
 opts.Add("CFLAGS", "Custom flags for the C compiler")
 opts.Add("CXXFLAGS", "Custom flags for the C++ compiler")
@@ -507,6 +508,10 @@ if selected_platform in platform_list:
         env.extra_suffix += "." + env["extra_suffix"]
 
     # Environment flags
+    CPPDEFINES = env.get("CPPDEFINES", "")
+    env["CPPDEFINES"] = ""
+    env.Append(CPPDEFINES=str(CPPDEFINES).split())
+
     CCFLAGS = env.get("CCFLAGS", "")
     env["CCFLAGS"] = ""
     env.Append(CCFLAGS=str(CCFLAGS).split())
```

That being said, I'm not overly concerned if we use different language here and in upstream Godot as this is a fairly niche repo that only a few users will ever need to compile. But if we can keep it in sync, it's a nice bonus.

Reply:
That being said (x2), the way `CCFLAGS` arguments are handled in upstream Godot (and my above diff) is broken for passing more than one argument...

Reply:
So, I wouldn't handle `CPPDEFINES` in the same way as upstream is handling `CCFLAGS` (by clearing them and then adding them all), because `CPPDEFINES` is calculated by scons internally (using prefixes and suffixes and other logic), it's not the same as generic variables that multiple compilers/build systems support. `CCFLAGS` and other flags of that type are wiped because they might be inherited in the environment from other compilers or from the IDE, which could confuse our build system. That's just not something that `CPPDEFINES` suffers from, `CPPDEFINES` is a scons-specific variable. That's why I'm merely adding them at the end in this PR, not doing a full wipe.

Reply:
Good point. So, considering this approach to adding defines is better, the consistency argument now reverts: it's the main Godot's `SConstruct` the one that may want to incorporate it and this PR is fine as it is.

Reply:
See https://github.com/godotengine/godot/pull/86964 - the upstream options were a mess, I think it's cleaner after this PR.

Reply:
I've updated this to match the change in https://github.com/godotengine/godot/pull/86964